src/dmd/dtemplate.d

```diff
@@ -2992,6 +2991,27 @@ void functionResolve(ref MatchAccumulator m, Dsymbol dstart, Loc loc, Scope* sc,
                 }
                 else
                     continue;   // MATCH.nomatch
+
+                // need to check here whether the constructor is the member of a struct
+                // declaration that defines a copy constructor. This is already checked
+                // in the semantic of CtorDeclaration, however, when matching functions,
+                // the template instance is not expanded.
+                auto ad = fd.isThis();
+                auto sd = ad.isStructDeclaration();
+                immutable dim = tf.parameterList.length;
+                if (sd && sd.hasCopyCtor && (dim == 1 || (dim > 1 && tf.parameterList[1].defaultArg)) )
```

Comment:
I see that the almost the same condition appears in two other places:
* https://github.com/dlang/dmd/blob/v2.099.0/src/dmd/dsymbolsem.d#L4048-L4052
* https://github.com/dlang/dmd/blob/v2.099.0/src/dmd/dsymbolsem.d#L4067-L4074

Can you refactor the code by adding a function that better clarifies the intent?

Reply:
Note that the difference between the 2 occurrences you provided is that one checks for an rvalue constructor, whereas the other one checks for a copy constructor. I would conflate the two.

Reply:
What I can do is to make the rvalue constructor check a function and look in a subsequent PR to create a function `isSpecialConstructor` which receives a TypeFunction and returns an enum value from `enum CtorType { Normal, CopyConstructor, RvalueConstructor}`